kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+static const char *realm;
+static const char *hdb;
+static const char *primary_server;
+static const char *admin_server;
+static const char *primary_admin_server;
+static const char *stash_file;
+static const char *kadmin_client_name = "keytab_service";
+static const char *kadmin_client_keytab;
+
+static krb5_error_code
+get_kadm_handle(krb5_context context, void **kadm_handle, int primary)
+{
+    kadm5_config_params conf;
+    krb5_error_code ret = 0;
+    const char *pstr;
+    const char *server = NULL;
+    char *s = NULL;
+
+    /* Configure kadmin connection */
+    memset(&conf, 0, sizeof(conf));
+    if (primary && local_read_only && !primary_admin_server)
+        return KADM5_READ_ONLY;
+    if (primary || !admin_server)
+        server = primary_admin_server;
+    else
+        server = admin_server;
+    if (server && (pstr = strrchr(server, ':'))) {
+        int32_t n;
+        char *ends;
+
+        if ((s = strdup(server)) == NULL)
+            return krb5_enomem(context);
+
+        errno = 0;
+        s[pstr - server] = '\0';
+        if ((n = strtol(++pstr, &ends, 10)) < 0 ||
+            n > UINT16_MAX || errno || *ends != '\0') {
+            errno = errno ? errno : ERANGE;
+            err(1, "Port number invalid in: %s", server);
+            free(s);
+            return ERANGE;
+        }
+        conf.kadmind_port = htons(n);
+        conf.mask |= KADM5_CONFIG_KADMIND_PORT;
+        server = s;
+    }
+    if (server) {
+        conf.admin_server = s ? strdup(server) : strdup(server);
+        conf.mask |= KADM5_CONFIG_ADMIN_SERVER;
+        free(s);
+        s = NULL;
+
+        if (conf.admin_server == NULL)
+            err(1, "Out of memory");
+    } else if (hdb) {
+        conf.dbname = strdup(hdb);
+        conf.mask |= KADM5_CONFIG_DBNAME;
+        if (conf.dbname == NULL)
+            err(1, "Out of memory");
+    }
+    if (realm) {
+        krb5_set_default_realm(context, realm); /* XXX ??? */
+        conf.realm = strdup(realm);
+        conf.mask |= KADM5_CONFIG_REALM;
+        if (conf.stash_file == NULL)
+            err(1, "Out of memory");
+    }
+    if (stash_file) {
+        conf.stash_file = strdup(stash_file);
+        conf.mask |= KADM5_CONFIG_KADMIND_PORT;
+        if (conf.stash_file == NULL)
+            err(1, "Out of memory");
+    }
+
+    if (!server) {
+        /* Local */
+        free(s);
+        return kadm5_s_init_with_password_ctx(context,
+                                              kadmin_client_name,
+                                              NULL,
+                                              NULL,
+                                              &conf, 0, 0,
+                                              kadm_handle);
+    }
+
+    /* Remote */
+    ret = kadm5_c_init_with_skey_ctx(context,
+                                     kadmin_client_name,
+                                     kadmin_client_keytab,
+                                     KADM5_ADMIN_SERVICE,
```

Comment:
Speaking of kadmin, we should forbid extracting keys with this service for any of the Heimdal KDC services like kadmin and iprop.